randovania/server/bot.py

```diff
@@ -53,16 +56,652 @@ async def reply_for_preset(message: discord.Message, versioned_preset: Versioned
+                "time_timestamp":player.time_timestamp,
+                "vod_timestamp":player.vod_timestamp,
+            })
+
+        race = {
+            "name" : self.name,
+            "permalink" : self.permalink,
+            "players" : players
+        }
+
+        return json.dumps(race)
+
+    @staticmethod
+    def from_json(json_data):
+        race = AsyncRace(json_data["name"], json_data["permalink"])
+        for player_json in json_data["players"]:
+            player = AsyncRacePlayer(player_json["username"])
+            player.state = async_race_state_from_string(player_json["state"])
+            player.time = player_json["time"]
+            player.vod_url = player_json["vod_url"]
+            player.play_timestamp = player_json["play_timestamp"]
+            player.time_timestamp = player_json["time_timestamp"]
+            player.vod_timestamp = player_json["vod_timestamp"]
+            race.players.append(player)
+        return race
+
+class AsyncRacePlayer:
+    username = ""
+    state = AsyncRacerState.NOT_STARTED
+    vod_url = ""
+    play_timestamp = 0
+    time_timestamp = 0
+    vod_timestamp = 0
+    time = 0 # race time in seconds
+
+    def __init__(self, username):
+        self.username = username
+        self.state = AsyncRacerState.NOT_STARTED
+        self.play_timestamp = 0
+        self.time_timestamp = 0
+        self.vod_timestamp = 0
+        self.vod_url = ""
+        self.time = 0
+
+    def reset(self):
+        self.__init__(self.username)
+
+async_races = list()
+
+def async_race_idx_from_name(name):
+    i = 0
+    for race in async_races:
+        if race.name.lower() == name.lower():
+            return i
+        i = i + 1
+    return None
+
+def async_player_idx_from_name(idx, name):
+    name = name.lower()
+    i = 0
+    for player in async_races[idx].players:
+        if player.username.lower() == name:
+            return i
+        i = i + 1
+    return None
+
+def async_race_usage(is_async_race_admin):
+    if is_async_race_admin:
+        return ASYNC_RACE_USAGE_ADMIN
+    return ASYNC_RACE_USAGE
+
+def async_race_get_active_player(player_name):
+    race_idx = 0
+    for race in async_races:
+        player_idx = 0
+        for player in race.players:
+            if player.username == player_name and (player.state == AsyncRacerState.RACING or player.state == AsyncRacerState.SUBMITTED_TIME):
+                return (race_idx, player_idx)
+            player_idx = player_idx + 1
+        race_idx = race_idx + 1
+    return (None, None)
+
+def async_race_string_to_time(string):
+    try:
+        string = string.split(":")
+        return int(string[0])*60*60 + int(string[1])*60 + int(string[2])
+    except Exception:
+        return None
+
+def format_seconds_to_hhmmss(seconds):
+    hours = seconds // (60*60)
+    seconds %= (60*60)
+    minutes = seconds // 60
+    seconds %= 60
+    return "%02i:%02i:%02i" % (hours, minutes, seconds)
```

Comment:
`str(datetime.timedelta(seconds=seconds))` gives precisely this result, but I dislike how there's no formatting arguments for this.
otherwise, I just found out `divmod` is a thing.